Reduces the amount of times total_moles() and heat_capacity() procs get called by replacing instances that expects the same result with variables that gets the result once instead.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Replaces various instances that call heat_capacity() and total_moles() with variables that only need to call them once instead.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Reduces the amount of times those procs get called. Probably slightly cheaper to get a value of a variable than to call a proc to get the same value.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
code: Reduces redundant total_moles() and heat_capacity() calls.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Unnecessary proc call reduction commit.

Reduces the amount of times total_moles() and heat_capacity() procs get called by replacing instances that expect the same result with variables that get the result once instead.

Pickle-Coding commented:
Boo merge conflict around 1 second after opening PR.

Pickle-Coding committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into small-stuff

LemonInTheDark commented:
To be clear, the benefits of storing it in a var are for readability and consistency. We're never gonna be in a position where the extra total moles call in hfr code is expensive. Careful to not run yourself ragged chasing ghosts yeah?

This is good tho

LemonInTheDark commented on code/modules/atmospherics/machinery/components/fusion/hfr_main_processes.dm:
```diff
@@ -406,29 +406,33 @@
 /obj/machinery/atmospherics/components/unary/hypertorus/core/proc/evaporate_moderator(delta_time)
+	//Moderator total moles
+	var/moderator_internal_total_moles = moderator_internal.total_moles()
 	// Don't evaporate if the reaction is dead
 	if (!power_level)
 		return
 	// All gases in the moderator slowly burn away over time, whether used for production or not
-	if(moderator_internal.total_moles() > 0)
-		moderator_internal.remove(moderator_internal.total_moles() * (1 - (1 - 0.0005 * power_level) ** delta_time))
+	if(moderator_internal_total_moles > 0)
+		moderator_internal.remove(moderator_internal_total_moles * (1 - (1 - 0.0005 * power_level) ** delta_time))
```
rather then doing this conversion, is there a reason this isn't just remove ratio? the exponent?

Pickle-Coding replied:
The expression on the right would converge to 1 as delta time increases, so I have no idea why it doesn't use remove ratio.

LemonInTheDark commented on code/modules/atmospherics/machinery/components/fusion/hfr_main_processes.dm:
```diff
@@ -505,39 +509,52 @@
 		var/fusion_temperature_delta = internal_fusion.temperature - moderator_internal.temperature
-		var/fusion_heat_amount = (1 - (1 - METALLIC_VOID_CONDUCTIVITY) ** delta_time) * fusion_temperature_delta * (internal_fusion.heat_capacity() * moderator_internal.heat_capacity() / (internal_fusion.heat_capacity() + moderator_internal.heat_capacity()))
-		internal_fusion.temperature = max(internal_fusion.temperature - fusion_heat_amount / internal_fusion.heat_capacity(), TCMB)
-		moderator_internal.temperature = max(moderator_internal.temperature + fusion_heat_amount / moderator_internal.heat_capacity(), TCMB)
+		var/fusion_heat_amount = (1 - (1 - METALLIC_VOID_CONDUCTIVITY) ** delta_time) * fusion_temperature_delta * (internal_fusion_heat_capacity * moderator_internal_heat_capacity / (internal_fusion_heat_capacity + moderator_internal_heat_capacity))
+		internal_fusion.temperature = max(internal_fusion.temperature - fusion_heat_amount / internal_fusion_heat_capacity, TCMB)
+		moderator_internal.temperature = max(moderator_internal.temperature + fusion_heat_amount / moderator_internal_heat_capacity, TCMB)
 
-	if(airs[1].total_moles() * 0.05 <= MINIMUM_MOLE_COUNT)
+	if(cooling_port_total_moles * 0.05 <= MINIMUM_MOLE_COUNT)
 		return
 	var/datum/gas_mixture/cooling_port = airs[1]
-	var/datum/gas_mixture/cooling_remove = cooling_port.remove(0.05 * cooling_port.total_moles())
+	var/datum/gas_mixture/cooling_removed = cooling_port.remove(0.05 * cooling_port_total_moles)
+	cooling_removed_heat_capacity = cooling_removed.heat_capacity()
 	//Cooling of the moderator gases with the cooling loop in and out the core
-	if(moderator_internal.total_moles() > 0)
-		var/coolant_temperature_delta = cooling_remove.temperature - moderator_internal.temperature
-		var/cooling_heat_amount = (1 - (1 - HIGH_EFFICIENCY_CONDUCTIVITY) ** delta_time) * coolant_temperature_delta * (cooling_remove.heat_capacity() * moderator_internal.heat_capacity() / (cooling_remove.heat_capacity() + moderator_internal.heat_capacity()))
-		cooling_remove.temperature = max(cooling_remove.temperature - cooling_heat_amount / cooling_remove.heat_capacity(), TCMB)
-		moderator_internal.temperature = max(moderator_internal.temperature + cooling_heat_amount / moderator_internal.heat_capacity(), TCMB)
-
-	else if(internal_fusion.total_moles() > 0)
-		var/coolant_temperature_delta = cooling_remove.temperature - internal_fusion.temperature
-		var/cooling_heat_amount = (1 - (1 - METALLIC_VOID_CONDUCTIVITY) ** delta_time) * coolant_temperature_delta * (cooling_remove.heat_capacity() * internal_fusion.heat_capacity() / (cooling_remove.heat_capacity() + internal_fusion.heat_capacity()))
-		cooling_remove.temperature = max(cooling_remove.temperature - cooling_heat_amount / cooling_remove.heat_capacity(), TCMB)
-		internal_fusion.temperature = max(internal_fusion.temperature + cooling_heat_amount / internal_fusion.heat_capacity(), TCMB)
-	cooling_port.merge(cooling_remove)
+	if(moderator_internal_total_moles > 0)
+		var/coolant_temperature_delta = cooling_removed.temperature - moderator_internal.temperature
+		var/cooling_heat_amount = (1 - (1 - HIGH_EFFICIENCY_CONDUCTIVITY) ** delta_time) * coolant_temperature_delta * (cooling_removed_heat_capacity * moderator_internal_heat_capacity / (cooling_removed_heat_capacity + moderator_internal_heat_capacity))
+		cooling_removed.temperature = max(cooling_removed.temperature - cooling_heat_amount / cooling_removed_heat_capacity, TCMB)
+		moderator_internal.temperature = max(moderator_internal.temperature + cooling_heat_amount / moderator_internal_heat_capacity, TCMB)
+
+	else if(internal_fusion_total_moles > 0)
+		var/coolant_temperature_delta = cooling_removed.temperature - internal_fusion.temperature
+		var/cooling_heat_amount = (1 - (1 - METALLIC_VOID_CONDUCTIVITY) ** delta_time) * coolant_temperature_delta * (cooling_removed_heat_capacity * internal_fusion_heat_capacity / (cooling_removed_heat_capacity + internal_fusion_heat_capacity))
+		cooling_removed.temperature = max(cooling_removed.temperature - cooling_heat_amount / cooling_removed_heat_capacity, TCMB)
+		internal_fusion.temperature = max(internal_fusion.temperature + cooling_heat_amount / internal_fusion_heat_capacity, TCMB)
+	cooling_port.merge(cooling_removed)
```
Bit uncomfortable with this. Using total_moles in this case isn't a huge deal, and when you remove it like this you open yourself up to some messy errors if the total changes after the proc. Would prefer this pattern of caching these values be kept to hot code and stuff that's simple

Pickle-Coding replied:
Why would the total changing after the proc cause errors?

LemonInTheDark replied:
It's not appropriate to cache the total moles of a gasmix when you want the total both before and after it "mutates", or changes.
Because of that, overusing this pattern can get a bit messy.
Also uh, `cooling_port_total_moles` doesn't need to exist, cause of remove_ratio

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/bluespace_sender.dm:
```diff
@@ -170,8 +170,9 @@
-				var/datum/gas_mixture/remove = bluespace_network.remove(bluespace_network.total_moles())
+			var/total_moles = bluespace_network.total_moles()
+			if(total_moles > 0)
+				var/datum/gas_mixture/remove = bluespace_network.remove(total_moles)
```
also should just be remove ratio (1), no need to cache

Pickle-Coding replied:
Not sure how I missed that.

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -192,7 +193,7 @@
 	else //Just siphoning all air
 
-		var/transfer_moles = environment.total_moles() * (volume_rate / environment.volume)
+		var/transfer_moles = total_moles * (volume_rate / environment.volume)
```
is there a point in caching it here? since these are in two different if blocks

Pickle-Coding replied:
Now that I think about it, no.

LemonInTheDark commented on code/modules/mob/living/carbon/alien/life.dm:
```diff
@@ -15,13 +16,13 @@
 
 	//Partial pressure of the plasma in our breath
-	var/Plasma_pp = (breath_gases[/datum/gas/plasma][MOLES]/breath.total_moles())*breath_pressure
+	var/Plasma_pp = (breath_gases[/datum/gas/plasma][MOLES] / total_moles) * breath_pressure
```
similar concern about mutation here

Pickle-Coding replied:
Mutation?

LemonInTheDark commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -237,28 +239,28 @@
 	//MIASMA
 	if(breath_gases[/datum/gas/miasma])
-		var/miasma_partialpressure = (breath_gases[/datum/gas/miasma][MOLES]/breath.total_moles())*breath_pressure
+		var/miasma_partialpressure = (breath_gases[/datum/gas/miasma][MOLES] / breath_total_moles) * breath_pressure
```
and here in lung code

Pickle-Coding commented:
I'll resolve these tomorrow (or later today depending on how you define when a day starts).

Pickle-Coding committed:
Requested changes.

Undoes the changes in life.dm in carbon and alien because of scary mutation stuff. Undoes changes in the process_internal_cooling() proc due to scary mutation stuff, and instead reduces redundant math by calling remote_ratio() and remove_specific_ratio(), instead of remove() and remove_specific() that multiplies by total moles.

Pickle-Coding committed:
t

Pickle-Coding commented:
FUCK

Pickle-Coding commented:
My git-fu isn't great, and I may have made a mistake. I guess I will have to restart this PR from scratch.
